smart-hadoop-support/smart-hadoop/src/main/java/org/smartdata/hdfs/scheduler/SmallFileScheduler.java

```diff
@@ -0,0 +1,512 @@
+        return false;
+      } else {
+        for (String smallFile : smallFileList) {
+          if (containerFileCache.contains(smallFile)) {
+            LOG.debug(smallFile + " is container file.");
+            return false;
+          }
+          if (compactSmallFileLock.contains(smallFile)) {
+            LOG.debug(smallFile + " is locked.");
+            return false;
+          }
+        }
+      }
+
+      return true;
+    } else if (UNCOMPACT_ACTION_NAME.equals(actionInfo.getActionName())) {
+      // Check if container file is not null and exist
+      String containerFilePath = actionInfo.getArgs().get(
+          SmallFileUncompactAction.CONTAINER_FILE);
+      try {
+        if (containerFilePath == null
+            || containerFilePath.isEmpty()
+            || !dfsClient.exists(containerFilePath)) {
+          LOG.debug("Illegal container file path: " + containerFilePath);
+          return false;
+        }
+      } catch (IOException e) {
+        LOG.error("Failed to check if container file exists: " + containerFilePath);
+        return false;
+      }
+
+      return true;
+    } else {
+      return true;
+    }
+  }
+
+  @Override
+  public ScheduleResult onSchedule(ActionInfo actionInfo, LaunchAction action) {
+    if (COMPACT_ACTION_NAME.equals(actionInfo.getActionName())) {
+      // Get container file and small file list of this action
+      String containerFilePath = actionInfo.getArgs().get(
+          SmallFileCompactAction.CONTAINER_FILE);
+      String smallFiles = actionInfo.getArgs().get(HdfsAction.FILE_PATH);
+      ArrayList<String> smallFileList = new Gson().fromJson(
+          smallFiles, new TypeToken<ArrayList<String>>() {
+          }.getType());
+
+      // Check if container file is locked and retry
+      if (containerFileLock.containsKey(containerFilePath)) {
+        return ScheduleResult.RETRY;
+      } else {
+        // Check if small file path is valid and unlocked
+        for (String smallFile : smallFileList) {
+          if (smallFile == null || smallFile.isEmpty()) {
+            LOG.debug("Illegal small file path: " + smallFile);
+            actionInfo.setResult("Illegal small file path: " + smallFile);
+            return ScheduleResult.FAIL;
+          } else if (compactSmallFileLock.contains(smallFile)) {
+            String errMsg = String.format("%s is locked.", smallFile);
+            LOG.debug(errMsg);
+            actionInfo.setResult(errMsg);
+            return ScheduleResult.FAIL;
+          } else if (containerFileCache.contains(smallFile)) {
+            String errMsg = String.format("%s is not small file.", smallFile);
+            LOG.debug(errMsg);
+            actionInfo.setResult(errMsg);
+            return ScheduleResult.FAIL;
+          }
+        }
+
+        // Get small file state map from meta store.
+        Map<String, FileState> fileStateMap;
+        try {
+          fileStateMap = metaStore.getFileStates(smallFileList);
```

Comment:
Need log for small files for the stage after been removed from small file lock and before been updated in metastore.